ruby/tests/message_test.rb

```diff
@@ -0,0 +1,214 @@
+      "repeated_int64"=>[],
+      "repeated_msg"=>[],
+      "repeated_string"=>["foo", "bar"],
+      "repeated_uint32"=>[],
+      "repeated_uint64"=>[],
+      "map_string_int32"=>{"bar"=>-123, "foo"=>123},
+      "map_string_msg"=>{"a"=>{"foo"=>1}}
+    }
+    assert_equal expected_value_hash, m.to_h
+  end
+
+
+  def test_to_h
+    m = TestMsg::Message.new
+    m.optional_string = 'foo'
+    m.repeated_string << 'bar'
+    m.repeated_uint32 << 10
+    assert_equal m.to_hash, m.to_h
+  end
+
+
+  def test_to_json
+    m = TestMsg::Message.new :optional_string => 'foo',
+                             :repeated_string => ['bar'],
+                             :repeated_bool   => [true, false],
+                             :repeated_uint32 => [10],
+                             :map_string_msg  => {"a"=>TestMsg::Message2.new(:foo => 1)}
+
+    result = JSON.parse(m.to_json)
+    expected_result = {
+      "map_string_int32"=>{},
+      "map_string_msg"=>{"a"=>{"foo"=>1}},
+      "optional_string"=>"foo",
+      "repeated_bool"=>[true, false],
+      "repeated_bytes"=>[],
+      "repeated_double"=>[],
+      "repeated_enum"=>[],
+      "repeated_float"=>[],
+      "repeated_int32"=>[],
+      "repeated_int64"=>[],
+      "repeated_msg"=>[],
+      "repeated_string"=>["bar"],
+      "repeated_uint32"=>[10],
+      "repeated_uint64"=>[]
+    }
+    if RUBY_PLATFORM == "java"
+      # FIXME: java has the correct result; the underlying C++ encode_json
+      # seems to be broken to some degree, where jruby returns the default values
+      # but the C library does not
+      expected_result = {
+        "optional_int32"=>0,
+        "optional_int64"=>0,
+        "optional_uint32"=>0,
+        "optional_uint64"=>0,
+        "optional_bool"=>false,
+        "optional_float"=>0.0,
+        "optional_double"=>0.0,
+        "optional_string"=>"foo",
+        "optional_bytes"=>"",
+        "optional_msg"=>nil,
+        "optional_enum"=>"Default",
+        "repeated_int32"=>[],
+        "repeated_int64"=>[],
+        "repeated_uint32"=>[10],
+        "repeated_uint64"=>[],
+        "repeated_bool"=>[true, false],
+        "repeated_float"=>[],
+        "repeated_double"=>[],
+        "repeated_string"=>["bar"],
+        "repeated_bytes"=>[],
+        "repeated_msg"=>[],
+        "repeated_enum"=>[],
+        "map_string_int32"=>{},
+        "map_string_msg"=>{"a"=>{"foo"=>1}}
+      }
+    end
+    assert_equal expected_result, result
+  end
+
+
+  def test_to_proto
+    m = TestMsg::Message.new :optional_string => 'foo',
+                             :repeated_string => ['bar'],
+                             :repeated_bool   => [true, false],
+                             :repeated_uint32 => [10],
+                             :map_string_msg  => {"a"=>TestMsg::Message2.new(:foo => 1)}
+
+    result = m.to_proto
+    expected_result = Base64.decode64("QgNmb29wCoABAYABAJoBA2JhcsIBBwoBYRICCAE=")
+    assert_equal expected_result, result
```

Comment:
**BUG** mri and jruby return different values.  I'm not sure which one is right.  The Base64 encoded value here is from the MRI version.  @cfallin thoughts?